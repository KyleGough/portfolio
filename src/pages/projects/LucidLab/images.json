[
  { "src": "/img/lucidlab1.jpg", "caption": "Upload test configuration page" },
  { "src": "/img/lucidlab4.jpg", "caption": "Upload IoT controller page" },
  { "src": "/img/lucidlab5.jpg", "caption": "Upload image page" },
  { "src": "/img/lucidlab2.jpg", "caption": "Mote RSSI Map" },
  { "src": "/img/lucidlab3.jpg", "caption": "Mote CCA Charts" }
]
