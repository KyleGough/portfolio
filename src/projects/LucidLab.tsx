import React, { useEffect } from 'react';
import { Section } from '../components/Section';
import { Divider } from '../components/Divider';
import { Chip } from '../components/Chip';
import { Pagination } from '../components/Pagination';
import { ImageFigure } from '../components/ImageFigure';
import { Screenshots } from '../components/Screenshots';

export default function LucidLab() {
  useEffect(() => {
    document.title =
      'LucidLab - An Adaptable, Heterogeneous Internet-of-things Testbed';
  }, []);

  const images = [
    { src: '/img/lucidlab1.jpg', caption: 'Upload test configuration page' },
    { src: '/img/lucidlab4.jpg', caption: 'Upload IoT controller page' },
    { src: '/img/lucidlab5.jpg', caption: 'Upload image page' },
    { src: '/img/lucidlab2.jpg', caption: 'Mote RSSI Map' },
    { src: '/img/lucidlab3.jpg', caption: 'Mote CCA Charts' },
  ];

  return (
    <>
      <Section>
        <h1 className="project-title">LucidLab</h1>
        <h2 className="project-subtitle">Master&apos;s Group Project</h2>
        <p className="text-link-hover my-4">
          <time dateTime="2019-10">October 2019</time> -{' '}
          <time dateTime="2020-05">May 2020</time>
        </p>
        <p className="mb-4 max-w-reading">
          An adaptable, heterogeneous IoT testbed for which I developed the
          front-end user interface. The UI allows registered users to upload and
          configure tests and images to be deployed on the testbed.
          Additionally, test results and custom metrics including mote
          availability graphs and mote CCA charts can be viewed.
        </p>

        <div className="flex flex-row flex-wrap items-center mt-8 gap-4">
          <Chip name="React" />
          <Chip name="JavaScript" />
          <Chip name="Materialize" />
          <Chip name="HTML" />
          <Chip name="CSS" />
          <Chip disabled name="Python" />
          <Chip disabled name="Shell" />
          <Chip disabled name="C" />
          <Chip disabled name="SQL" />
        </div>
      </Section>

      <Divider />

      <ImageFigure image={images[0]} />

      <Divider />

      <Section>
        <h2 className="project-header">Abstract</h2>
        <p className="mb-4 max-w-reading">
          With the advent of smarter, low-power wireless sensor hardware (or
          “motes”) along with the influx of Internet of Things (IoT) devices,
          researchers are leveraging these technologies to form the networks
          underpinning “smart” homes and cities. Of vital importance in this
          process are wireless sensor network (WSN) testbeds, which allow
          researchers to debug the firmware that runs on motes. A particularly
          challenging exercise is presented when combining heterogeneous
          wireless sensor hardware and proprietary IoT devices on a single
          network, enabling development of applications that interact with both
          motes and IoT devices. Whilst existing WSN testbeds permit the use of
          a variety of sensor hardware, they do not allow interaction with
          off-the-shelf IoT devices, which already have a strong presence in
          many households.
        </p>
        <p className="mb-4 max-w-reading">
          This report presents LucidLab, a WSN testbed designed to directly
          integrate IoT devices, facilitating the debugging of richer
          applications. In addition to the standard features of a WSN testbed,
          such as logging of output through a serial connection, LucidLab
          provides an adaptable architecture through simple and scalable
          addition of hardware. IoT integration is enabled using openHAB as an
          IoT substrate.
        </p>
        <p className="mb-4 max-w-reading">
          Users can control IoT devices on the network through scripts uploaded
          alongside firmware that interpret mote output to manage IoT devices
          accordingly. By virtue of LucidLab’s prompt logging process, mote
          output is quickly passed to the user-specified script, allowing
          responsive actuation of IoT devices. The proposed implementation
          provides a practical system combining IoT and WSN devices whilst
          offering a simple and intuitive deployment process, thereby enabling
          rapid testing of the applications forming the backbone of “smart”homes
          and cities
        </p>
      </Section>

      <Divider />

      <Screenshots images={images} />

      <Divider />

      <Pagination
        previousTitle="Logical Sudoku Solver"
        previousLink="/projects/sudoku"
        nextTitle="Portfolio"
        nextLink="/projects/portfolio"
      />
    </>
  );
}
