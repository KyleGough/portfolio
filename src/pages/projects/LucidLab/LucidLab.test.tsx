import { render } from '@testing-library/react';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';

import { LucidLab } from './LucidLab';

describe('LucidLab project page', () => {
  it('renders', async () => {
    const { container } = render(
      <BrowserRouter>
        <LucidLab />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });
});
