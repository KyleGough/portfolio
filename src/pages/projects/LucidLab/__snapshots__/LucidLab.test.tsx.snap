// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LucidLab project page renders 1`] = `
<div>
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <h1
      class="project-title"
    >
      LucidLab
    </h1>
    <h2
      class="project-subtitle"
    >
      Master's Group Project
    </h2>
    <p
      class="text-link-hover my-4"
    >
      <time
        datetime="2019-10"
      >
        October 2019
      </time>
       
      -
       
      <time
        datetime="2020-05"
      >
        May 2020
      </time>
    </p>
    <p
      class="mb-4 max-w-reading"
    >
      An adaptable, heterogeneous IoT testbed for which I developed the front-end user interface. The UI allows registered users to upload and configure tests and images to be deployed on the testbed. Additionally, test results and custom metrics including mote availability graphs and mote CCA charts can be viewed.
    </p>
    <div
      class="flex flex-row flex-wrap items-center mt-8 gap-4"
    >
      <div
        class="text-chip bg-chip-light text-xs px-4 py-0.5 h-8 leading-chip rounded-2xl font-extrabold"
      >
        React
      </div>
      <div
        class="text-chip bg-chip-light text-xs px-4 py-0.5 h-8 leading-chip rounded-2xl font-extrabold"
      >
        JavaScript
      </div>
      <div
        class="text-chip bg-chip-light text-xs px-4 py-0.5 h-8 leading-chip rounded-2xl font-extrabold"
      >
        Materialize
      </div>
      <div
        class="text-chip bg-chip-light text-xs px-4 py-0.5 h-8 leading-chip rounded-2xl font-extrabold"
      >
        HTML
      </div>
      <div
        class="text-chip bg-chip-light text-xs px-4 py-0.5 h-8 leading-chip rounded-2xl font-extrabold"
      >
        CSS
      </div>
      <div
        class="text-disabled bg-chip-old text-xs px-4 py-0.5 h-8 leading-chip rounded-2xl font-extrabold"
      >
        Python
      </div>
      <div
        class="text-disabled bg-chip-old text-xs px-4 py-0.5 h-8 leading-chip rounded-2xl font-extrabold"
      >
        Shell
      </div>
      <div
        class="text-disabled bg-chip-old text-xs px-4 py-0.5 h-8 leading-chip rounded-2xl font-extrabold"
      >
        C
      </div>
      <div
        class="text-disabled bg-chip-old text-xs px-4 py-0.5 h-8 leading-chip rounded-2xl font-extrabold"
      >
        SQL
      </div>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <img
      alt="Upload test configuration page"
      class="border-2 border-white rounded-lg text-center items-center shadow mx-auto"
      src="/img/lucidlab1.jpg"
    />
    <p
      class="mt-4 mb-2 text-sm text-center text-light font-thin"
    >
      Upload test configuration page
    </p>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <h2
      class="project-header"
    >
      Abstract
    </h2>
    <p
      class="mb-4 max-w-reading"
    >
      With the advent of smarter, low-power wireless sensor hardware (or “motes”) along with the influx of Internet of Things (IoT) devices, researchers are leveraging these technologies to form the networks underpinning “smart” homes and cities. Of vital importance in this process are wireless sensor network (WSN) testbeds, which allow researchers to debug the firmware that runs on motes. A particularly challenging exercise is presented when combining heterogeneous wireless sensor hardware and proprietary IoT devices on a single network, enabling development of applications that interact with both motes and IoT devices. Whilst existing WSN testbeds permit the use of a variety of sensor hardware, they do not allow interaction with off-the-shelf IoT devices, which already have a strong presence in many households.
    </p>
    <p
      class="mb-4 max-w-reading"
    >
      This report presents LucidLab, a WSN testbed designed to directly integrate IoT devices, facilitating the debugging of richer applications. In addition to the standard features of a WSN testbed, such as logging of output through a serial connection, LucidLab provides an adaptable architecture through simple and scalable addition of hardware. IoT integration is enabled using openHAB as an IoT substrate.
    </p>
    <p
      class="mb-4 max-w-reading"
    >
      Users can control IoT devices on the network through scripts uploaded alongside firmware that interpret mote output to manage IoT devices accordingly. By virtue of LucidLab’s prompt logging process, mote output is quickly passed to the user-specified script, allowing responsive actuation of IoT devices. The proposed implementation provides a practical system combining IoT and WSN devices whilst offering a simple and intuitive deployment process, thereby enabling rapid testing of the applications forming the backbone of “smart”homes and cities
    </p>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <h2
      class="text-3xl mb-12"
    >
      Screenshots
    </h2>
    <div
      class="grid grid-cols-12 gap-4"
    >
      <figure
        class="col-span-12 md:col-span-6 lg:col-span-4"
      >
        <img
          alt="Upload test configuration page"
          class="cursor-pointer border-2 border-link hover:border-link-hover"
          loading="lazy"
          src="/img/lucidlab1.jpg"
        />
        <figcaption
          class="my-2 text-sm text-center text-light font-thin"
        >
          Upload test configuration page
        </figcaption>
      </figure>
      <figure
        class="col-span-12 md:col-span-6 lg:col-span-4"
      >
        <img
          alt="Upload IoT controller page"
          class="cursor-pointer border-2 border-link hover:border-link-hover"
          loading="lazy"
          src="/img/lucidlab4.jpg"
        />
        <figcaption
          class="my-2 text-sm text-center text-light font-thin"
        >
          Upload IoT controller page
        </figcaption>
      </figure>
      <figure
        class="col-span-12 md:col-span-6 lg:col-span-4"
      >
        <img
          alt="Upload image page"
          class="cursor-pointer border-2 border-link hover:border-link-hover"
          loading="lazy"
          src="/img/lucidlab5.jpg"
        />
        <figcaption
          class="my-2 text-sm text-center text-light font-thin"
        >
          Upload image page
        </figcaption>
      </figure>
      <figure
        class="col-span-12 md:col-span-6 lg:col-span-4"
      >
        <img
          alt="Mote RSSI Map"
          class="cursor-pointer border-2 border-link hover:border-link-hover"
          loading="lazy"
          src="/img/lucidlab2.jpg"
        />
        <figcaption
          class="my-2 text-sm text-center text-light font-thin"
        >
          Mote RSSI Map
        </figcaption>
      </figure>
      <figure
        class="col-span-12 md:col-span-6 lg:col-span-4"
      >
        <img
          alt="Mote CCA Charts"
          class="cursor-pointer border-2 border-link hover:border-link-hover"
          loading="lazy"
          src="/img/lucidlab3.jpg"
        />
        <figcaption
          class="my-2 text-sm text-center text-light font-thin"
        >
          Mote CCA Charts
        </figcaption>
      </figure>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="flex justify-between"
    >
      <a
        class="flex items-center fill-link text-link hover:fill-link-hover hover:text-link-hover"
        href="/projects/sudoku"
      >
        <svg
          class="w-6 h-6"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
            opacity=".87"
          />
          <path
            d="M17.51 3.87L15.73 2.1 5.84 12l9.9 9.9 1.77-1.77L9.38 12l8.13-8.13z"
          />
        </svg>
        <div
          class="text-center ml-4"
        >
          <p
            class="text-sm"
          >
            Previous Project
          </p>
          <p>
            Logical Sudoku Solver
          </p>
        </div>
      </a>
      <a
        class="flex items-center fill-link text-link hover:fill-link-hover hover:text-link-hover"
        href="/projects/portfolio"
      >
        <div
          class="text-center mr-4"
        >
          <p
            class="text-sm"
          >
            Next Project
          </p>
          <p>
            Portfolio
          </p>
        </div>
        <svg
          class="w-6 h-6"
          enable-background="new 0 0 24 24"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M0,0h24v24H0V0z"
              fill="none"
            />
          </g>
          <g>
            <polygon
              points="6.23,20.23 8,22 18,12 8,2 6.23,3.77 14.46,12"
            />
          </g>
        </svg>
      </a>
    </div>
  </div>
</div>
`;
